source/slang/syntax.cpp

```diff
@@ -940,6 +940,155 @@ void Type::accept(IValVisitor* visitor, void* extra)
+            // search for a substitution that might apply to us
+            for (auto s = subst; s; s = s->outer.Ptr())
+            {
+                // the generic decl associated with the substitution list must be
+                // the generic decl that declared this parameter
+                auto genericDecl = s->genericDecl;
+                if (genericDecl != parentDeclRef.getDecl()->ParentDecl)
+                    continue;
+                int index = 0;
+                for (auto m : genericDecl->Members)
+                {
+                    if (m.Ptr() == parentDeclRef.getDecl())
+                    {
+                        // We've found it, so return the corresponding specialization argument
+                        (*ioDiff)++;
+                        if (auto declRef = s->args[index].As<DeclRefType>())
+                        {
+                            newParentDeclRef = (*declRef).declRef.As<AggTypeDecl>();
+                            goto searchEnd;
+                        }
+                    }
+                    else if (auto typeParam = m.As<GenericTypeParamDecl>())
+                    {
+                        index++;
+                    }
+                    else if (auto valParam = m.As<GenericValueParamDecl>())
+                    {
+                        index++;
+                    }
+                    else
+                    {
+                    }
+                }
+            }
+        searchEnd:
+            if (newParentDeclRef)
+            {
+                Decl* targetTypeDecl = nullptr;
+                if (newParentDeclRef.getDecl()->memberDictionary.TryGetValue(this->GetDeclRef().decl->getName(), targetTypeDecl))
+                {
+                    if (auto typeDefDecl = targetTypeDecl->As<TypeDefDecl>())
+                        return GetType(DeclRef<TypeDefDecl>(typeDefDecl, subst));
+                    else
+                        return DeclRefType::Create(this->getSession(), DeclRef<Decl>(targetTypeDecl, subst));
+                }
+            }
+        }
+       
+        return this;
+    }
+
+    int AssocTypeDeclRefType::GetHashCode()
+    {
+        return declRef.GetHashCode();
+    }
+
+    Type* AssocTypeDeclRefType::CreateCanonicalType()
+    {
+        return this;
+    }
+
+    // GenericConstraintDeclRefType
+
+    String GenericConstraintDeclRefType::ToString()
+    {
+        // TODO: what is appropriate here?
+        return "<GenericConstraintType>";
+    }
+
+    bool GenericConstraintDeclRefType::EqualsImpl(Type * type)
+    {
+        if (auto other = type->As<GenericConstraintDeclRefType>())
+        {
+            return supType->Equals(other->supType) && subType->Equals(other->subType);
+        }
+        return false;
+    }
+
+    RefPtr<Val> GenericConstraintDeclRefType::SubstituteImpl(Substitutions* subst, int* ioDiff)
+    {
+        auto genParamDecl = subType.As<DeclRefType>()->declRef.As<GenericTypeParamDecl>();
```

Comment:
Wait. We are just applying substitution to the `sub` type here. I'm thoroughly confused about what this type is supposed to represent now.

The original rationale I had around the code that you changed was that given something like:

```
interface IFoo { void foo(); }
__generic<T : IFoo> void bar( T t ) { t.foo(); }
```

In the expression `t.foo` the front-end should elaborate this into something like `t.{ T : IFoo }.{IFoo::foo}`. That is, it first looks up the "member" representing the conformance of `T` to `IFoo` on the particular value `t` (so this is a `MemberExpr`). If we were just referencing the constraint `{T : IFoo}` directly, that would be a `SubtypeWitness` value, but because we are accessing the member on `t`, it made sense to me to treat the expression `t.{T:IFoo}` as having type `IFoo` - that is, we are conceptually "casting" `t` to `IFoo`, by looking up its member that represents the inheritance (similar to how you could translate C++ `struct D : B {};` into C `struct D { B base; };`).

Then we look up the `IFoo::foo` member on the resulting expression, which works just fine because the base expression now has type `IFoo`. It works just fine for ordinary interface stuff.

With your changes, the `t.{T : IFoo}` lookup now results in a `GenericConstraintDeclRefType(T, IFoo)`. If that is just a kind of proxy type that effectively behaves like an `IFoo` for the purposes of lookup/codegen/etc. then maybe that's okay.

But then when you do substitution of `T => C` for some concrete `C`, you go from `GenericConstraintDeclRefType(T,IFoo)` to just `C`, which seems wrong. The right answer would seem to be to create the appropriate `GenericConstraintDeclRefType(C,IFoo)`, which should encode a reference to the `{C : IFoo}` inheritance decl, which witnesses the conformance.

But that's the place where I think the existing approach was already set up to do the 100% right thing. We already had `t.{T : IFoo}` where the `{T : IFoo}` is a `DeclRef` to the appropriate generic parameter type constraint. The arguments to the generic should then have `{C : IFoo}` at the appropriate argument position, and we should be able to substitute that in for the `{T : IFoo}` so that we have `t.{C:IFoo}` which is effectively exactly what we want.

It seems like the real problem here might be that not all the code paths around generics are properly doing the Right Thing and treating generic type constraints like explicit parameters/arguments, and so you've had to make this whole side-band path to get the info through.